feat(browser): Session replay GA (#16499)
newrelic/docs-website: src/content/docs/browser/browser-monitoring/browser-pro-features/session-replay.mdx

import browserErrorsInboxSessionReplay from 'images/browser_screenshot-full_errors-inbox-session-replay.webp'

import browserViewConsumptionDetails from 'images/browser_screenshot-crop_view-consumption-details.webp'

<Callout variant="important">

> Session replay consumption will be billable starting May 8, 2024. See [Data consumption](#view-consumption) for details.

Author: ```suggestion
Session replay will be billable starting May 8, 2024. See [Data consumption](#view-consumption) for details.
```

import browserErrorsInboxSessionReplay from 'images/browser_screenshot-full_errors-inbox-session-replay.webp'

import browserViewConsumptionDetails from 'images/browser_screenshot-crop_view-consumption-details.webp'

<Callout variant="important">
Session replay consumption will be billable starting May 8, 2024. See [Data consumption](#view-consumption) for details.
</Callout>

> Session replay allows you to play back user interactions on your web app and map other telemetry data to help you understand your user's journeys. Session replay can be used for troubleshooting and improving the end-user experience, such as:

Reviewer: Look for places where you can remove filler phrases and words like "allows you to". 

I'd replace it with something like "Session replay plays back user interactions ... and maps other telemetry data... etc" By removing those fillers, it often leads to stronger, more straightforward sentences.

Reviewer: But otherwise, I this this is a very strong intro! I love the bullets with the concrete use cases and the way the animated GIF peeks up from below the fold. Well done!

import browserErrorsInboxSessionReplay from 'images/browser_screenshot-full_errors-inbox-session-replay.webp'

<Callout variant="important">

> Session replay consumption will be billable starting May 1, 2024.

Author: ```suggestion
Session replay consumption will be billable starting May 8, 2024.
```

>  Session replay becomes a paid feature on May 1, 2024. Try it free now!

Author: ```suggestion
 Session replay consumption will be billable starting May 1, 2024.
```

* <DoNotTranslate>**Improving the user experience**</DoNotTranslate>: Discover areas of your web app that are causing users frustration. For example, you might see that users are getting lost in your navigation or clicking on buttons that don't do anything.
* <DoNotTranslate>**Identifying issues in key user journeys**</DoNotTranslate>: Identify areas of your web app that are causing performance problems. For example, if session replay reveals the `/cart/checkout` page is experiencing excessive page load time, you'd prioritize fixing this critical user journey.

> Here's how a sample session replay looks:

Reviewer: Or maybe even: "Here's session replay in action:"?

<img
    title="Session replay details"
    alt="Screenshot of session replay details"
    src={browserSessionReplayDetails}
/>

## How session replay works [#how-it-works]

When a user loads your web page, session replay takes a snapshot of the Document Object Model (DOM), which represents the page's structure and visual elements. To ensure accurate styling information, the browser agent accesses your CSS files. The agent then integrates these styles directly into the replay recording, eliminating the need to reference external CSS files during playback. As the user interacts with the page, a library within the browser agent captures any changes made to the DOM. This captured data is then stored securely in New Relic's database. 

Replays store up to:
* 4 hours of a user session
* 30 minutes of inactivity

> Data retention depends on your account. By default, browser event data is stored 8 days, but DataPlus customers have 98 days, and others may have even longer retentions.

Reviewer: This "others may have even longer retentions" feels pretty odd and nebulous. Why mention it if we don't tell people how to get longer retention if they want it?

> Data retention depends on your account. By default, browser event data is stored 8 days, but DataPlus customers have 98 days, and others may have even longer retentions.

Author: Agreeed.

```suggestion
By default, browser event data is stored 8 days, but actual data retention depends on your account.
```
@CherylFrankenfield Would this revision work?

<Steps>

  <Step>
    ### Enable session replay [#enable-configure-settings]

>     As of April 1, 2024, session replay settings will be automatically enabled for all new browser apps. For apps createad before April 1, you'll need to manually enable the settings:

Author: Is this true for both session replay and session tracking settings?

<Steps>

  <Step>
    ### Enable session replay [#enable-configure-settings]

>     As of April 1, 2024, session replay settings will be automatically enabled for all new browser apps. For apps createad before April 1, you'll need to manually enable the settings:

Author: ```suggestion
```

    
Set a sampling rate (0%-100%) for the following samples:

    * <DoNotTranslate>**User sessions**</DoNotTranslate> records a random sample of all user sessions.

    * <DoNotTranslate>**Errored session**</DoNotTranslate> records a random sample of all errored sessions. The agent buffers for up to 30 seconds on the client-side, and if errors occur, the session will be recorded. If an error does not occur, the replay is not harvested. 

> Here's an example of configuring sampling rates: If you set the session sampling rate to 50% and the error sampling rate to 100%, it means that:

Reviewer: The doubled up colons feel odd to me

* Half of all user sessions will be recorded.
* All errored sessions will be recorded, even if they're not part of the 50% sample.

  </Step>
<Step>

> ### Configure privacy settings [#configure-privacy-settings]

Reviewer: After reading this section, one thing that's not clear to me: what are New Relic's default privacy protections? Are there any? Is this a critical step to complete before implementing session replay?

> <Callout variant="caution">

Reviewer: I'm wondering if we can give more guidance/clarity here. Are we saying that we believe New Relic's default privacy settings are what people should be using and people should check with their privacy teams before changing them?

<img
    title="Example of default privacy settings"
    alt="Example of default privacy settings"
    src={browserTextMaskingExample}
/>

If you need to change the default privacy settings, you can choose specific classes and attributes to:

* Mask user input
* Block site content
* Ignore user input

### Mask user input [#mask-input]

You can mask user input, which means the input will be replaced with asterisks (*). For example, a user's login information will render as asterisks in session replay.

To specify which input should be masked, you have a few options:

  * On the <DoNotTranslate>**Application settings**</DoNotTranslate> page, add your own CSS selectors or check the masking boxes. Note that using mask selectors will **not** mask user input, so if you need to hide user input, we recommend using block selectors. 

          <img
              title="Mask settings"
              alt="Screenshot of mask settings in the ui"
              src={browserMaskSettings}
          />
  * Add our CSS class `nr-mask` or attribute `data-nr-mask` to your webpage HTML.

>   For example, if you want to mask user-entered email address, add `nr-mask` to your email class attribute in `<div>`:

Reviewer 2: You cannot mask user-entered anything. You would use the User input masking options or the block selectors.

Author: @jomle What's an example of something (non-user input) customers would want to mask?

Author: @jomle Can you provide another example (the html sample below) of what they would want to mask?

      ```html
      <div class="email-input nr-mask">
        <input type="email" name="email" placeholder="Enter your email address">

>       </div>

Reviewer 2: This is not correct. You cannot use the mask selectors on <input> fields. An example for mask selectors would be like a <div> with an account ID in it.

Author: @jomle Do you mean "would be like a URL with an account ID in it"?

Reviewer 2: Oh I guess I typed an HTML DIV tag and it hid it. An anchor tag with a URL would also work.

### Block site content [#block-content]

You can block entire sections of content on your site, which means the section will appear as an empty placeholder in session replay. For example, if you have an image on your About Us webpage and don’t need it captured, you could block the class containing the image.

To block specific classes or attributes, you have a few options:

* On the Application settings page, add your CSS selectors in the **Block selectors* field.

      <img
          title="Block Selectors"
          alt="Screenshot of session replay settings"
          src={browserSessionReplayAppSettings}
      />
*  Manually add our CSS class `nr-block` or attribute `data-nr-block` to your webpage HTML. For example, if you wanted to block irrelevant text and images from session replay, add `nr-block` to `<div class>`:

    ```html
    <html>
    <head>
      <title>Sample image and text</title>
    </head>
    <body>
      <div class="image-text-container nr-block">
        <img src="https://example.com/image.png" alt="Image description">
      </div>
    </body>
    </html>
    ```

### Ignore user input [#ignore-input]

> You can ignore user input entirely, which means the input field won't show up at all in session replay. For example, if you want to ignore a user's email address and password, you could ignore the element that contains those fields.

Reviewer 2: Ignoring means you ignore changes to the input field. The input field will still be displayed in the replay, but you won't see changes to the value.

> You can ignore user input entirely, which means the input field won't show up at all in session replay. For example, if you want to ignore a user's email address and password, you could ignore the element that contains those fields.

Author: ```suggestion
You can ignore changes to a user input field, which means the input field will still be displayed in the replay, but you won't display changes to the value. For example, you could ignore the class containing the email address and password fields.
```

  >
      If you want to customize exactly what is masked with asterisks, you have a few options:

>         * On the <DoNotTranslate>**Application settings**</DoNotTranslate> page, add your own CSS selectors or check the masking boxes. Note that using mask selectors will **not** mask user input, so if you need to hide user input, we recommend using block selectors. 

Author: Is this correct?

>         * On the <DoNotTranslate>**Application settings**</DoNotTranslate> page, add your own CSS selectors or check the masking boxes. Note that using mask selectors will **not** mask user input, so if you need to hide user input, we recommend using block selectors. 

Author: @CherylFrankenfield Can you confirm this description is correct?

Reviewer 2: Masking selectors do not affect user input. This is correct.

  </Collapser>
  </CollapserGroup>

> * <DoNotTranslate>**Block**</DoNotTranslate> means content will be replaced with an empty placeholder. For example, if you have an image on your About Us webpage and don’t need it captured, you could block the class containing the image.

Author: Is a screenshot the best example for the Block feature? Is there another one more relevant to customers?

  </Collapser>
  </CollapserGroup>

> * <DoNotTranslate>**Block**</DoNotTranslate> means content will be replaced with an empty placeholder. For example, if you have an image on your About Us webpage and don’t need it captured, you could block the class containing the image.

Author: ```suggestion
* <DoNotTranslate>**Block**</DoNotTranslate> means content will be replaced with an empty placeholder. For example, if you have text fields for sensitive billing information, you could block the class containing those fields.
```

> <CollapserGroup>

Reviewer: I'm inclined to want to move the troubleshooting section to the browser monitoring troubleshooting category (the doc is already so long!) I think it will make it easier for people to find the information in these two collapsers.

https://deploy-preview-16499--docs-website-netlify.netlify.app/docs/browser/new-relic-browser/troubleshooting/troubleshoot-ajax-data-collection/

* Session replay works by rebuilding the user's DOM, which includes your website's structure and visual elements. To achieve this, our agent needs to access your CSS files to accurately capture styling information. We then inline these styles directly into the replay recording, eliminating the need to reference the original CSS files during playback.

* If your CSS files are hosted on a separate domain, CORS may prevent our agent from accessing them.

> * The agent attempts to collect all CSS, but if it can't, it sends the replay with broken styles along with attributes `hasCrossOriginCSS:true` and `inlinedCrossOriginCSS:false` so the session replay UI knows some styles are broken and provides a warning message with a link to a fix. The agent is set up to report:
>     * `invalidStylesheetsDetected` will resolve to true if we find cross-origin stylesheets
>     * `inlinedAllStylesheets` will resolve to false if any stylesheets weren’t able to be inlined by the agent. If false, you'll see a message in the UI to fix it.

Author: ```suggestion
```

</Steps>

## (Optional) Manually record session replays [#manual-replays]

You can manually invoke a session to be collected by using these two API methods:

* [`NREM.startSessionReplay`](/docs/browser/new-relic-browser/browser-apis/start/)
* [`NREM.stopSessionReplay`](/docs/browser/new-relic-browser/browser-apis/stop/)

If you use these methods, you'll need to set both sampling rates in application settings to 0%. This is beneficial if you’d like a user to opt in to having user interactions recorded for page performance and observability purposes.  

> 

Author: @metal-messiah @CherylFrankenfield I think this section needs more info, so here are some questions I have:

1. When it says "both sampling rates", should this just say "all sampling rates" to include the new session traces one?
2. The sentence is about "This is beneficial if you'd like a user to opt in to having interactions recorded" is confusing. What does this mean? Why would a manual recording be any different than a standard sampling,  AKA what's the use case for using this manual process?
3. How exactly does this work? As a reader, if I follow the docs links out to the API start/stop docs, it's not really helpful/doesn't tell me how it's applicable for session replay, so I think any relevant content from the other docs should be implemented/contextualized here.

* Session replay works by rebuilding the user's DOM, which includes your website's structure and visual elements. To achieve this, our agent needs to access your CSS files to accurately capture styling information. We then inline these styles directly into the replay recording, eliminating the need to reference the original CSS files during playback.

* If your CSS files are hosted on a separate domain, CORS may prevent our agent from accessing them.

* The agent attempts to collect all CSS, but if it can't, it sends the replay with broken styles along with attributes `hasCrossOriginCSS:true` and `inlinedCrossOriginCSS:false` so the session replay UI knows some styles are broken and provides a warning message with a link to a fix. The agent is set up to report:
    * `invalidStylesheetsDetected` will resolve to true if we find cross-origin stylesheets

>     * `inlinedAllStylesheets` will resolve to false if any stylesheets weren’t able to be inlined by the agent. If false, you'll see a message in the UI to fix it.

Reviewer 2: The SessionReplay events are not visible to customers so I don't know why we would tell them attributes on the event.

Author: Thanks for this callout. @CherylFrankenfield , if customers cannot see sessionreplay events, I think I should remove these bullet points. Do you have any objections?

                    "enabled": true,
                    "collectFonts": true,
                    "blockSelector": ".private-fields",
                    "maskTextSelector": "*",
                    "inlineStylesheet": false,
                    "maskAllInputs": true,
                    "maskInputOptions": {
                      "color": true,
                      "date": true,
                      "textArea": true
                    },
                    "samplingRate": 9.99,
                    "errorSamplingRate": 1.11
                  }
                }
                ```
            </td>
        </tr>
    </tbody>
</table>

## Data consumption [#data-consumption]

### View consumption [#view-consumption]

Between April 8 and May 8, 2024, while session replay bytes are free, you can view your consumption to help with planning once the bytes become paid:
    1. Go to <DoNotTranslate>**Application Settings**</DoNotTranslate>.
    2. Scroll down to <DoNotTranslate>**Session replay**</DoNotTranslate>.
    3. In the <DoNotTranslate>**Session sample size**</DoNotTranslate> section, click <DoNotTranslate>**View your account's consumption**</DoNotTranslate>.

>     <img
>         title="View account consumption on application settings page"
>         alt="screenshot of application settings page"
>         src={browserViewConsumptionDetails}
>     />

Author: ```suggestion
    <img
        title="View account consumption on application settings page"
        alt="screenshot of application settings page"
        src={browserViewConsumptionDetails}
        style={{ maxWidth: '50%' }}
    />
```

Here's an example of adding the `crossorigin="anonymous"` attribute:

```
<link rel="stylesheet" href="assets.yoursite.com/styles.css" crossorigin="anonymous">
```
</Collapser>
</CollapserGroup>

## Manually record session replays [#manual-replays]

> If needed, you can programmatically invoke a session replay using these two API methods:

Reviewer: (Outside of the scope of this PR) This feels like an edge use case and a good candidate for moving to its own doc, I think. Also, it would be great to have more context about why someone would want to do this (but maybe we don't know yet?)

* [`NREM.startSessionReplay`](/docs/browser/new-relic-browser/browser-apis/start/)
* [`NREM.stopSessionReplay`](/docs/browser/new-relic-browser/browser-apis/stop/)

Here are some use cases:

* Allow users to opt in for interaction recording. For example, if you have a pop-up asking "Allow this session to be recorded for performance analysis?":
    1. Upon user consent, call `NREM.startSessionReplay` to initiate recording. 
    2. Once the session is complete or consent is withdrawn, use `NREM.stopSessionReplay` to stop capturing data.
* Record sessions based on specific criteria. For example:
    * Start recording only for sessions on specific URLs such as a product checkout flow.
    * Stop recording for sensitive pages such as `/billing` or `/health-records`.

To ensure all user interactions are captured during manual recording, set sampling rates to 0% in your application settings.

> ## Use Nerdgraph to configure replay settings [#nerdgraph-mutation]

Reviewer: I kind of want to move this info to the Nerdgraph section of the docs.

Author: I think I'll include this in a post-release PR. Good idea.

                        collectFonts
                        inlineImages
                        inlineStylesheet
                        samplingRate
                        maskAllInputs
                        maskInputOptions {
                          color
                          date
                          datetimeLocal
                          email
                          month
                          number
                          range
                          search
                          select
                          tel
                          text
                          textArea
                          time
                          url
                          week
                        }
                      }
                    }
                    errors {
                      field
                      errorClass
                      description
                    }
                  }
                }
                ```
            </td>
            <td>
                ```json
                {
                  "sessionReplay": {
                    "enabled": true,
                    "collectFonts": true,

>                     "blockSelector": "String",

Reviewer 2: Nobody would use "String" as their blockSelector. Try ".private-fields"

>                     "maskTextSelector": "String",

Reviewer 2: Nobody would use "String" as their maskTextSelector. Let's use "*" (which is the default masking option)

                    "maskInputOptions": {
                      "color": true,
                      "date": true,
                      "textArea": true
                    },
                    "samplingRate": 9.99,
                    "errorSamplingRate": 1.11
                  }
                }
                ```
            </td>
        </tr>
    </tbody>
</table>

## Data consumption [#data-consumption]

### View consumption [#view-consumption]

Between April 8 and May 8, 2024, while session replay bytes are free, you can view your consumption to help with planning once the bytes become paid:
    1. Go to <DoNotTranslate>**Application Settings**</DoNotTranslate>.
    2. Scroll down to <DoNotTranslate>**Session replay**</DoNotTranslate>.
    3. In the <DoNotTranslate>**Session sample size**</DoNotTranslate> section, click <DoNotTranslate>**View your account's consumption**</DoNotTranslate>.

    <img
        title="View account consumption on application settings page"
        alt="screenshot of application settings page"
        src={browserViewConsumptionDetails}
    />

### Plan for paid consumption [#plan-for-paid-consumption]

> After May 8, 2024, session replay will follow the same consumption pricing as your other browser bytes. The amount of bytes produced depends on the count, length, and user-activity levels of sessions, as well as the complexity of your site's DOM.

Reviewer: I assume there's a plan to remove the April and May dates after we pass May 8?

Author: Yes, I've got a followup ticket for this!

The most accurate way to project your cost per replay is to enable the feature for a short test period and measure your actual consumption. Alternatively, you can use an average of 5.3 MB per replay as a rough approximation, though keep in mind that your actual cost may vary significantly. A simple formula to project your approximate costs is: `sessions` x `sampling_rate` x `gb_per_replay` x `cost_per_gb`
        * Example: `sessions: 1 million` x `sampling_rate: 5%` x `gb_per_replay: 0.0053` x `cost_per_gb: $0.30` = `$79.50 for 50,000 replays`

To control your consumption, [adjust your sampling rates](#configure-sampling-rates).
    

> ## User privacy and security concerns [#data-security]

Reviewer: I understand not wanting to bulk up the steps section of the doc, but this seems like super useful/important context for the configure privacy settings step.

* Session replay works by rebuilding the user's DOM, which includes your website's structure and visual elements. To achieve this, our agent needs to access your CSS files to accurately capture styling information. We then inline these styles directly into the replay recording, eliminating the need to reference the original CSS files during playback.

* If your CSS files are hosted on a separate domain, CORS may prevent our agent from accessing them. The agent attempts to fix the broken CSS. If it cannot, it sends the replay with broken styles along with attributes `hasCrossOriginCSS:true` and `inlinedCrossOriginCSS:false` so the UI knows some styles are broken and we can notify the user watching it with his to fix it. The agent is set up to report:
  * `inlinedAllStylesheets`: if it was successful at inlining or not
  * invalidStylesheetsDetected: if stylesheets were detected that could not be read by JS

> This can result in incomplete or inaccurate replays, as essential styling information may be missing.

Author: Placeholder. This section will need to be revised.

* Session replay works by rebuilding the user's DOM, which includes your website's structure and visual elements. To achieve this, our agent needs to access your CSS files to accurately capture styling information. We then inline these styles directly into the replay recording, eliminating the need to reference the original CSS files during playback.

> * If your CSS files are hosted on a separate domain, CORS may prevent our agent from accessing them. The agent attempts to fix the broken CSS. If it cannot, it sends the replay with broken styles along with attributes `hasCrossOriginCSS:true` and `inlinedCrossOriginCSS:false` so the UI knows some styles are broken and we can notify the user watching it with his to fix it. The agent is set up to report:

Author: @CherylFrankenfield @metal-messiah How exactly does the agent attempt to fix the broken css? How does it appear in the UI ("...so the UI knows some styles are broken and we can notify users of how to fix it...")?

Author: @CherylFrankenfield What does the UI link "See how to fix it" link to?

  * `inlinedAllStylesheets`: if it was successful at inlining or not

>   * invalidStylesheetsDetected: if stylesheets were detected that could not be read by the JavaScript

Author: ```suggestion
  * `invalidStylesheetsDetected`: if stylesheets were detected that could not be read by the JavaScript
```

```
</Collapser>
</CollapserGroup>

## Use Nerdgraph to configure replay settings [#nerdgraph-mutation]

> You can configure session replay settings using our [NerdGraph API](/docs/apis/nerdgraph/get-started/introduction-new-relic-nerdgraph/) instead of using the UI. This method is helpful when...

Author: @CherylFrankenfield When would customers need/want to use nerdgraph instead of the UI?

> You can configure session replay settings using our [NerdGraph API](/docs/apis/nerdgraph/get-started/introduction-new-relic-nerdgraph/) instead of using the UI. This method is helpful when...

Author: ```suggestion
You can configure session replay settings using our [NerdGraph API](/docs/apis/nerdgraph/get-started/introduction-new-relic-nerdgraph/) instead of using the UI. This method is helpful if you need to make bulk changes, such as applying sampling rates to multiple browser entities.
```
